### ISSUE TITLE
Instance for Data.Complex

### ISSUE THREAD

**Issue by chessai** [haskell/primitive#176]
It seems like there could be an instance for `Complex a`, so long as `a` can unbox. I just wrote this for some FFT stuff I was doing using `MutablePrimArray`:

```haskell
instance Prim e => Prim (Complex e) where
  sizeOf# _ = 2# *# sizeOf# (undefined :: e)
  alignment# _ = alignment# (undefined :: e)
  indexByteArray# arr# i# =
    let i = I# i#
        arr = ByteArray arr#
    in
         (indexByteArray arr (2 * i + 0))
      :+ (indexByteArray arr (2 * i + 1))
  readByteArray# :: forall s e. (Prim e) => MutableByteArray# s -> Int# -> State# s -> (# State# s, Complex e #)
  readByteArray# arr# i# = internal $ do
    let i = I# i#
        arr = MutableByteArray arr#
    a <- readByteArray arr (2 * i + 0) :: ST s e
    b <- readByteArray arr (2 * i + 1)
    return (a :+ b)
  writeByteArray# :: forall s e. (Prim e) => MutableByteArray# s -> Int# -> Complex e -> State# s -> State# s
  writeByteArray# arr# i# (a :+ b) = internal_ $ do
    let i = I# i#
        arr = MutableByteArray arr#
    writeByteArray arr (2 * i + 0) a
    writeByteArray arr (2 * i + 1) b :: ST s ()
  setByteArray# = defaultSetByteArray#
  indexOffAddr# :: Addr# -> Int# -> Complex e
  indexOffAddr# addr# i# =
    let i = I# i#
        addr = Addr addr#
    in (indexOffAddr addr (2 * i + 0)) :+ (indexOffAddr addr (2 * i + 1))
  readOffAddr# :: forall s e. (Prim e) => Addr# -> Int# -> State# s -> (# State# s, Complex e #)
  readOffAddr# addr# i# = internal $ do
    let i = I# i#
        addr = Addr addr#
    a <- readOffAddr addr (2 * i + 0) :: ST s e
    b <- readOffAddr addr (2 * i + 1)
    return (a :+ b)
  writeOffAddr# :: forall s e. (Prim e) => Addr# -> Int# -> Complex e -> State# s -> State# s
  writeOffAddr# addr# i# (a :+ b) = internal_ $ do
    let i = I# i#
        addr = Addr addr#
    writeOffAddr addr (2 * i + 0) a
    writeOffAddr addr (2 * i + 1) b :: ST s ()
  setOffAddr# = defaultSetOffAddr#

internal_ :: PrimBase m => m () -> State# (PrimState m) -> State# (PrimState m)
internal_ m s = case internal m s of
  (# s', _ #) -> s'

```

**Comment by treeowl** [comment]
I suppose this is probably reasonable. The components of a `Complex` have the same type, so sticking them together shouldn't lead to alignment issues, I don't think. But the whole alignment business is terribly under-documented.

**Comment by andrewthad** [comment]
This is related to https://github.com/haskell/primitive/issues/70. The driving question behind both: Do we admit `Prim` instances for things that are not directly backed by a single `Int#`/`Word#`. I'm inclined to say yes. I want to be able to stick `Complex` in a `PrimArray` just like I want to be able to stick `Fingerprint` in a `PrimArray`. I feel similarly about the pass-thru newtype wrappers from `base` like `Identity`, `Down`, `First`, `Last`, and `Dual`.

Back to the issue at hand. I agree with David that this does not cause any alignment issues. We probably want to write the typeclass methods out with primitives instead of boxed types, since that's what this library does for everything else. But other than that, I'm good with this.

**Comment by chessai** [comment]
@treeowl What about alignment needs to be documented?

Also, can I make a PR adding this/any appropriate tests?

**Comment by cartazio** [comment]
@chessai what are our semantics for alignment at all, for "builtin" vs user types

**Comment by cartazio** [comment]
but yeah, Complex sounds kosher

one issue that i only just realized we need to articulate/balance is whether these "aggregate"/composite instances are SOA or AOS memory layouts

in Vector, Storable is AOS, and Unboxed is SOA, but both layouts are GOOD approaches, and we shouldn't preclude them

**Comment by andrewthad** [comment]
Anything I've ever done with the `Prim` typeclass is AOS. To do SOA well, you really want separate arrays. The `vector` library can handle this with `Unboxed` because there we have the flexibility to pick out what vector-like data type we're using, so we can pick a pair/triple/quadruple of arrays. But with `Prim`, you have to use a single array, and the typeclass doesn't have as many methods that we can give efficient implementations to.

And I just now realized that all the range-copying operations in `Data.Primitive.ByteArray`/`Data.Primitive.PrimArray` would be totally broken for SOA. For example, `copyPrimArray` would not work correctly since its implementation copies a single, contiguous range.

Basically, what I'm trying to say is that the design of the `Prim` typeclass precludes SOA. I think it's worth pointing that out in the docs, directing people with that need to `vector`, and providing AOS `Prim` instances for types where there aren't alignment problems with doing so (meaning, still not doing any kind of horrendous `instance (Prim a, Prim b) => Prim (a,b)` nonsense).

**Comment by chessai** [comment]
Where should these instances be located? AFAIK I need to have access to the lifted prim functions (dothingByteArray / dothingAddr), which are not accessible in Data.Primitive.Types

**Comment by andrewthad** [comment]
You can write the instance for `Complex` without using those functions. The docs for [`defaultSetByteArray#`](http://hackage.haskell.org/package/primitive-0.6.4.0/docs/Data-Primitive-Types.html#v:defaultSetByteArray-35-) have part of an example of how to do this.

**Comment by chessai** [comment]
ah, makes sense

**Comment by konsumlamm** [comment]
This was implemented in #385.
